File: app/src/main/java/com/rendersoncs/report/view/login/LoginActivity.java
package com.rendersoncs.report.view.login;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;

import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FacebookAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.auth.TwitterAuthProvider;
import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.rendersoncs.report.R;
import com.rendersoncs.report.view.main.MainActivity;
import com.rendersoncs.report.view.login.util.User;

import java.util.Arrays;

public class LoginActivity extends CommonActivity {

    private static final int RC_SIGN_IN_GOOGLE = 7859;
    private static final String FACEBOOK = "facebook";
    private static final String GOOGLE = "google";
    private static final String GOOGLE_TOKEN = "940299608698-s73ntnnnlbauh7lm0pb8ouis82d279m7.apps.googleusercontent.com";

    private FirebaseAuth mAuth;
    private FirebaseAuth.AuthStateListener mAuthListener;

    private User user;
    private CallbackManager callbackManager;
    private GoogleSignInClient mGoogleApiClient;
    /*private GoogleApiClient mGoogleApiClient;
    private TwitterAuthClient twitterAuthClient;*/

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true);

        // FACEBOOK SIGN IN
        //FacebookSdk.sdkInitialize(getApplicationContext());
        callbackManager = CallbackManager.Factory.create();
        LoginManager.getInstance().registerCallback(callbackManager, new FacebookCallback<LoginResult>() {
            @Override
            public void onSuccess(LoginResult loginResult) {
                accessFacebookLoginData(loginResult.getAccessToken());
            }

            @Override
            public void onCancel() { }

            @Override
            public void onError(FacebookException error) {
                FirebaseCrashlytics.getInstance().recordException(error);
                closeProgressBar();
                showSnackBar(getResources().getString(R.string.label_login_facebook_failed));
            }
        });

        // GOOGLE SIGN IN
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(GOOGLE_TOKEN)
                .requestEmail()
                .build();

        mGoogleApiClient = GoogleSignIn.getClient(this, gso);
        /*mGoogleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this, this)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();*/

        // TWITTER
        /*twitterAuthClient = new TwitterAuthClient();*/

        mAuth = FirebaseAuth.getInstance();
        mAuthListener = getFirebaseAuthResultHandler();
        initViews();
        initUser();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RC_SIGN_IN_GOOGLE) {

            GoogleSignInResult googleSignInResult = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
            GoogleSignInAccount account = googleSignInResult.getSignInAccount();

            if (account == null) {
                showSnackBar(getResources().getString(R.string.label_google_failed));
                return;
            }

            accessGoogleLoginData(account.getIdToken());

        } else {
            //twitterAuthClient.onActivityResult(requestCode, resultCode, data);
            callbackManager.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        verifyLogged();
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mAuthListener != null) {
            mAuth.removeAuthStateListener(mAuthListener);
        }
    }

    // ACCESS FACEBOOK
    private void accessFacebookLoginData(AccessToken accessToken) {
        accessLoginData(
                FACEBOOK,
                (accessToken != null ? accessToken.getToken() : null)
        );
    }

    // ACCESS GOOGLE
    private void accessGoogleLoginData(String accessToken) {
        accessLoginData(
                GOOGLE,
                accessToken
        );
    }

    // ACCESS TWITTER
    /*private void accessTwitterLoginData(String token, String secret, String id) {
        accessLoginData(
                "twitter",
                token,
                secret
        );
    }*/

    private void accessLoginData(String provider, String... tokens) {
        if (tokens != null
                && tokens.length > 0
                && tokens[0] != null) {

            AuthCredential credential = FacebookAuthProvider.getCredential(tokens[0]);
            credential = provider.equalsIgnoreCase(GOOGLE) ? GoogleAuthProvider.getCredential(tokens[0], null) : credential;
            credential = provider.equalsIgnoreCase("twitter") ? TwitterAuthProvider.getCredential(tokens[0], tokens[1]) : credential;
            //credential = provider.equalsIgnoreCase("github") ? GithubAuthProvider.getCredential( tokens[0] ) : credential;

            user.saveProviderSP(LoginActivity.this, provider);
            mAuth.signInWithCredential(credential)
                    .addOnCompleteListener(task -> {
                        if (!task.isSuccessful()) {
                            showSnackBar(getResources().getString(R.string.label_login_social_failed));
                        }
                    })
                    .addOnFailureListener(Throwable::printStackTrace);
        } else {
            mAuth.signOut();
        }
    }

    private FirebaseAuth.AuthStateListener getFirebaseAuthResultHandler() {
        return (firebaseAuth -> {

            FirebaseUser userFirebase = firebaseAuth.getCurrentUser();

            if (userFirebase == null) {
                return;
            }

            if (user.getId() == null
                    && isNameOk(user, userFirebase)) {

                user.setId(userFirebase.getUid());
                user.setNameIfNull(userFirebase.getDisplayName());
                user.setEmailIfNull(userFirebase.getEmail());
                user.saveDB();
            }

            callMainActivity();
        });
    }

    private boolean isNameOk(User user, FirebaseUser firebaseUser) {
        return (
                user.getName() != null
                        || firebaseUser.getDisplayName() != null
        );
    }


    @Override
    protected void initViews() {
        email = findViewById(R.id.email);
        password = findViewById(R.id.password);
        progressBar = findViewById(R.id.login_progress);
    }

    @Override
    protected void initUser() {
        user = new User();
        user.setEmail(email.getText().toString());
        user.setPassword(password.getText().toString());
    }

    public void callSignUp(View view) {
        Intent intent = new Intent(this, SignUpActivity.class);
        startActivity(intent);
    }

    public void callReset(View view) {
        Intent intent = new Intent(this, RecoveryLoginActivity.class);
        startActivity(intent);
    }

    public void sendLoginData(View view) {
        if (email.getText().toString().isEmpty() || password.getText().toString().isEmpty()) {
            showSnackBar(getResources().getString(R.string.label_please_entry_data));
        } else {
            openProgressBar();
            initUser();
            verifyLogin();
        }
    }

    // login Facebook
    public void sendLoginFacebookData(View view) {
        /*Toast.makeText(getApplicationContext(), getResources().getString(R.string.version_beta), Toast.LENGTH_SHORT).show();*/
        LoginManager
                .getInstance()
                .logInWithReadPermissions(
                        this,
                        Arrays.asList("public_profile", "user_friends", "email")
                );
    }

    // login Twitter
    /*public void sendLoginTwitterData(View view) {
        FirebaseCrash.log("LoginActivity:clickListener:button:sendLoginTwitterData()");
        twitterAuthClient.authorize(
                this,
                new Callback<TwitterSession>() {
                    @Override
                    public void success(Result<TwitterSession> result) {

                        TwitterSession session = result.data;

                        accessTwitterLoginData(
                                session.getAuthToken().token,
                                session.getAuthToken().secret,
                                String.valueOf( session.getUserId() )
                        );
                    }
                    @Override
                    public void failure(TwitterException exception) {
                        FirebaseCrash.report( exception );
                        showSnackBar( exception.getMessage() );
                    }
                }
        );
    }*/

    // Google
    public void sendLoginGoogleData(View view) {
        /*Toast.makeText(getApplicationContext(), getResources().getString(R.string.version_beta), Toast.LENGTH_SHORT).show();
        Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient);*/
        Intent signInIntent = mGoogleApiClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN_GOOGLE);
    }


    private void verifyLogged() {
        if (mAuth.getCurrentUser() != null) {
            callMainActivity();
        } else {
            mAuth.addAuthStateListener(mAuthListener);
        }
    }

    private void verifyLogin() {
        user.saveProviderSP(LoginActivity.this, "");
        mAuth.signInWithEmailAndPassword(
                user.getEmail(),
                user.getPassword()
        )
                .addOnCompleteListener(task -> {

                    if (!task.isSuccessful()) {
                        closeProgressBar();
                    }
                }).addOnFailureListener(e -> {
            FirebaseCrashlytics.getInstance().recordException(e);
            showSnackBar(e.getMessage());
        });
    }

    private void callMainActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }

    // Google
    /*@Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        Crashlytics.logException(new Exception(connectionResult.getErrorCode() + ": " + connectionResult.getErrorMessage()));
        showSnackBar(connectionResult.getErrorMessage());
    }*/
}

File: app/src/main/java/com/rendersoncs/report/view/login/LoginActivity.kt
package com.rendersoncs.report.view.login

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.activity.viewModels
import com.facebook.AccessToken
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.*
import com.google.firebase.auth.FirebaseAuth.AuthStateListener
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.rendersoncs.report.R
import com.rendersoncs.report.common.constants.ReportConstants
import com.rendersoncs.report.databinding.FragmentLoginBinding
import com.rendersoncs.report.common.util.closeVirtualKeyBoard
import com.rendersoncs.report.view.login.util.User
import com.rendersoncs.report.view.login.viewmodel.LoginViewModel
import com.rendersoncs.report.view.main.MainActivity
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class LoginActivity : CommonActivity(), OnEditorActionListener {
    private var mAuth: FirebaseAuth? = null
    private var mAuthListener: AuthStateListener? = null
    private var user: User? = null
    private var callbackManager: CallbackManager? = null
    private var mGoogleApiClient: GoogleSignInClient? = null
    private lateinit var binding: FragmentLoginBinding

    private val viewModel: LoginViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = FragmentLoginBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        FirebaseCrashlytics.getInstance().isCrashlyticsCollectionEnabled = true

        // FACEBOOK SIGN IN
        callbackManager = CallbackManager.Factory.create()
        LoginManager.getInstance().registerCallback(callbackManager, object : FacebookCallback<LoginResult> {
            override fun onSuccess(result: LoginResult) {
                accessFacebookLoginData(result.accessToken)
            }

            override fun onCancel() {}
            override fun onError(error: FacebookException) {
                FirebaseCrashlytics.getInstance().recordException(error)
                closeProgressBar()
                showSnackBar(resources.getString(R.string.label_login_facebook_failed))
            }
        })

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(ReportConstants.GOOGLE.GOOGLE_TOKEN)
                .requestEmail()
                .build()
        mGoogleApiClient = GoogleSignIn.getClient(this, gso)
        mAuth = FirebaseAuth.getInstance()
        mAuthListener = firebaseAuthResultHandler
        setListener()
        initViews()
        initUser()
    }

    private fun setListener() = with(binding) {
        singIn.setOnClickListener {
            sendLoginData()
        }
        signUp.setOnClickListener {
            callSignUp()
        }
        recoveryPassword.setOnClickListener {
            callReset()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_SIGN_IN_GOOGLE && data != null) {
            val googleSignInResult = Auth.GoogleSignInApi.getSignInResultFromIntent(data)
            val account = googleSignInResult!!.signInAccount
            if (account == null) {
                showSnackBar(resources.getString(R.string.label_google_failed))
                return
            }
            accessGoogleLoginData(account.idToken)
        } else {
            callbackManager!!.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onStart() {
        super.onStart()
        verifyLogged()
    }

    override fun onStop() {
        super.onStop()
        mAuthListener?.let {
            mAuth?.removeAuthStateListener(it)
        }
    }

    private fun accessFacebookLoginData(accessToken: AccessToken?) {
        accessLoginData(
                FACEBOOK,
                accessToken?.token
        )
    }

    private fun accessGoogleLoginData(accessToken: String?) {
        accessLoginData(
                GOOGLE,
                accessToken
        )
    }

    private fun accessLoginData(provider: String, vararg tokens: String?) {
        if (tokens.isNotEmpty() && tokens[0] != null) {
            var credential: AuthCredential? = FacebookAuthProvider.getCredential(tokens[0] ?: "")
            credential = if (provider.equals(GOOGLE, ignoreCase = true)) GoogleAuthProvider.getCredential(tokens[0], null) else credential
            credential = if (provider.equals("twitter", ignoreCase = true)) TwitterAuthProvider.getCredential(tokens[0] ?: "", tokens[1] ?: "") else credential

            user?.saveProviderSP(this@LoginActivity, provider)
            mAuth?.signInWithCredential(credential!!)
                    ?.addOnCompleteListener { task: Task<AuthResult?> ->
                        if (!task.isSuccessful) {
                            showSnackBar(resources.getString(R.string.label_login_social_failed))
                        }
                    }
                    ?.addOnFailureListener { obj: Exception -> obj.printStackTrace() }
        } else {
            mAuth?.signOut()
        }
    }

    private val firebaseAuthResultHandler: AuthStateListener
        get() = AuthStateListener { firebaseAuth: FirebaseAuth ->
            val userFirebase = firebaseAuth.currentUser ?: return@AuthStateListener
            if (user?.id == null && isNameOk(user, userFirebase)) {
                user?.id = userFirebase.uid
                user?.setNameIfNull(userFirebase.displayName)
                user?.setEmailIfNull(userFirebase.email)
                user?.setUrlImgIfNull(userFirebase.photoUrl)
                user?.saveDB()
            }
            callMainActivity()
        }

    private fun isNameOk(user: User?, firebaseUser: FirebaseUser): Boolean {
        return (user?.name != null || firebaseUser.displayName != null)
    }

    override fun initViews() {
        progressBar = findViewById(R.id.login_progress)

        binding.email.addTextChangedListener(emailWatcher)
        binding.password.addTextChangedListener(passwordWatcher)
        binding.password.setOnEditorActionListener(this)
    }

    private val emailWatcher = object: TextWatcher {
        override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

        override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

        override fun afterTextChanged(s: Editable?) {
            validate(s, binding.textInputEmail, R.string.label_insert_email)
        }
    }

    private val passwordWatcher = object: TextWatcher {
        override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

        override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

        override fun afterTextChanged(s: Editable?) {
            validate(s, binding.textInputPassword, R.string.label_insert_password)
        }
    }

    override fun initUser() {
        user = User()
        user?.email = binding.email.text.toString()
        user?.password = binding.password.text.toString()
    }

    private fun callSignUp() {
        val intent = Intent(this, SignUpActivity::class.java)
        startActivity(intent)
    }

    private fun callReset() {
        val intent = Intent(this, RecoveryLoginActivity::class.java)
        startActivity(intent)
    }

    private fun sendLoginData() {
        when {
            binding.email.text.toString().isEmpty() -> {
                binding.textInputEmail.error = resources.getString(R.string.label_insert_email)
            }
            binding.password.text.toString().isEmpty() -> {
                binding.textInputPassword.error = resources.getString(R.string.label_insert_password)
            }
            else -> {
                openProgressBar()
                initUser()
                verifyLogin()
            }
        }
    }

    override fun onEditorAction(view: TextView?, actionId: Int, event: KeyEvent?): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            closeVirtualKeyBoard(this, view!!)
            sendLoginData()
            return true
        }
        return false
    }

    fun sendLoginFacebookData(view: View?) {
        LoginManager
                .getInstance()
                .logInWithReadPermissions(
                        this,
                        listOf("public_profile", "user_friends", "email")
                )
    }

    fun sendLoginGoogleData(view: View?) {
        val signInIntent = mGoogleApiClient!!.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN_GOOGLE)
    }

    private fun verifyLogged() {
        if (mAuth?.currentUser != null) {
            callMainActivity()
        } else {
            mAuth?.addAuthStateListener(mAuthListener!!)
        }
    }

    private fun verifyLogin() {
        user?.saveProviderSP(this@LoginActivity, "")
        mAuth?.signInWithEmailAndPassword(user?.email ?: "", user?.password ?: "")
            ?.addOnCompleteListener { task: Task<AuthResult?> ->
                if (!task.isSuccessful) {
                    closeProgressBar()
                }
                if (task.isComplete) {
                    mAuth?.currentUser?.let { currentUser ->
                        val user = com.rendersoncs.report.model.User(
                            userId = currentUser.uid
                        )
                        viewModel.insertUserBD(user)
                    }
                }
            }?.addOnFailureListener { e: Exception ->
                FirebaseCrashlytics.getInstance().recordException(e)
                showSnackBar(e.message)
            }
    }

    private fun callMainActivity() {
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        finish()
    }

    override fun onBackPressed() {
        finish()
        super.onBackPressed()
    }

    companion object {
        private const val RC_SIGN_IN_GOOGLE = 7859
        private const val FACEBOOK = "facebook"
        private const val GOOGLE = "google"
    }
}